server/src/main/java/org/elasticsearch/common/compress/DeflateCompressor.java

```diff
@@ -73,8 +74,75 @@ public int headerLength() {
 
+    // Reusable inflater reference for streaming decompression
+    private static final ThreadLocal<ReleasableReference<Inflater>> inflaterForStreamRef = ThreadLocal.withInitial(() -> {
+        final Inflater inflater = new Inflater(true);
+        return new ReleasableReference<>(inflater, inflater::reset);
+    });
+
+    // Reusable deflater reference for streaming compression
+    private static final ThreadLocal<ReleasableReference<Deflater>> deflaterForStreamRef = ThreadLocal.withInitial(() -> {
+        final Deflater deflater = new Deflater(LEVEL, true);
+        return new ReleasableReference<>(deflater, deflater::reset);
+    });
+
+    // Reference to a deflater or inflater that is used to make sure we do not use the same stream twice when nesting streams.
+    private static final class ReleasableReference<T> implements Releasable {
+
+        protected final T resource;
+
+        private final Releasable releasable;
+
+        // Thread that is currently using this reference. Only used for assertions and only assigned if assertions are enabled.
+        private Thread thread = null;
+
+        // true if this reference is currently in use and is not available for re-use
+        boolean inUse;
+
+        protected ReleasableReference(T resource, Releasable releasable) {
+            this.resource = resource;
+            this.releasable = releasable;
+        }
+
+        T get() {
+            if (Assertions.ENABLED) {
+                assert thread == null;
+                thread = Thread.currentThread();
+            }
+            assert inUse == false;
+            inUse = true;
+            return resource;
+        }
+
+        @Override
+        public void close() {
+            if (Assertions.ENABLED) {
+                assert thread == Thread.currentThread() :
+                        "Opened on [" + thread.getName() + "] but closed on [" + Thread.currentThread().getName() + "]";
+                thread = null;
+            }
+            assert inUse;
+            inUse = false;
+            releasable.close();
+        }
+    }
+
     @Override
-    public StreamInput streamInput(StreamInput in) throws IOException {
+    public StreamInput threadLocalStreamInput(StreamInput in) throws IOException {
+        return new InputStreamStreamInput(inputStream(in, true));
+    }
+
+    /**
+     * Creates a new input stream that decompresses the contents read from the provided input stream.
+     * Closing the returned stream will close the provided input stream.
+     * Optionally uses thread-local, pooled resources to save off-heap allocations if the stream is guaranteed to not escape the current
+     * thread.
+     *
+     * @param in           input stream to wrap
+     * @param threadLocal  whether this stream will only be used on the current thread or not
+     * @return             decompressing stream
+     */
+    public static InputStream inputStream(InputStream in, boolean threadLocal) throws IOException {
```

Comment:
I'm fine with this as a one-off special case.

> Also, I wonder if we shouldn't just make all the methods on this class static and do away with the Compressor interface. We only have one implementation that we use via its singleton, so there isn't much point in doing in having that interface around?

This is a good point; there was a time when both DEFLATE and LZF were available. I hesitate on removing the interface since there may be a desire to allow other compression algorithms in the future; I'm not aware of any specific effort but that doesn't mean there won't be one to investigate/add something else like Zstd.